Arm/State.lean

```diff
@@ -662,3 +668,327 @@ theorem write_mem_irrelevant :
   simp [read_mem, write_mem, store_write_irrelevant]
 
 end Memory
+
+
+/-!
+# New definitions for the memory model
+
+For freedom in experimenting with definitions,
+we define our own version of `read_mem` and `write_mem`,
+called `read_mem'` and `write_mem'`.
+These operate directly on the memory, rather than the `ArmState`.
+We prove their equivalence to the existing definitions
+(`read_mem_eq_read_mem'`, `write_mem_eq_write_mem'`).
+
+Furthermore, we define equivalences to `BitVec.extractLsByte`
+to ease reasoning about byte-level manipulation.
+As an easy corollary, we get `getLsb` theorems on these to allow
+`omega` based reasoning about bit-level values of memory.
+-/
+
+def Memory.read (addr : BitVec 64) (m : Memory) : BitVec 8 :=
+  read_store addr m
+
+theorem ArmState.read_mem_eq_mem_read : read_mem addr s = s.mem.read addr := rfl
+
+/--
+A variant of `write_mem` that directly talks about writes to memory, instead of over the entire `ArmState`
+-/
+def Memory.write (addr : BitVec 64) (val : BitVec 8) (m : Memory) : Memory :=
+  write_store addr val m
+
+theorem ArmState.write_mem_eq_mem_write :  (write_mem addr val s).mem = s.mem.write addr val := rfl
+
+namespace Memory
+
+/--
+A variant of `read_mem` that directly talks about writes to memory, instead of over the entire `ArmState`
+-/
+
+@[memory_simp_rules]
+theorem getLsb_read (mem : Memory) : (mem.read addr).getLsb i = (mem addr).getLsb i := rfl
```

Comment:
Why do we need this theorem? It feels like it's getting rid of the abstraction boundary, so it really should only be an internal lemma, but then it's part of a simp-set

Reply:
@alexkeizer `getLsb` is meant to be simplified away in the internal theory, but I'm happy to un-simp it if you like.